### DISCUSSION TITLE
Best practice to setup connection pool for Postgres (GCP Cloud SQL) and Vercel?

### DISCUSSION THREAD

**Discussion by teddis** [prisma/prisma#24497] [Q&A]
### Question

I'm deploying a Next app to Vercel with GCP Cloud SQL (Postgres v15). I'm observing horrendous performance accessing PG from a single cron job invocation that starts an edge function on Vercel (Node.js). Queries are extremely fast locally, under 50ms on a Mac. Obviously there's network overhead but it should not be this slow. I'm seeing each query take ~1.5s when they should be sub 200ms. My Cloud SQL instance is located in Ohio where my Cloud SQL is located. Assume a new connection is established on every job invocation. But even worse, this job lasts 5 minutes and should scream after the connection is setup.

What is the best practice to establish a connection pool and/or address this problem? How do I implement this, configure, monitor? Does Prisma have built-in pooling? Is there a GCP connection pool? Should my Next app manage the pool?

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by nurul3101** [comment]
Hello @teddis 👋 

In serverless environments like Vercel, each function invocation can result in a new connection to the database, which can quickly exhaust the database connection limit and lead to increased latency. So, I would definitely recommend a connection pooler. The recommend connection poolers are listed [here](https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/databases-connections#external-connection-poolers).

I would also recommend adding observability to get a better idea on where exactly is the time spent for queries experiencing higher latencies. You could use [metrics](https://www.prisma.io/docs/orm/prisma-client/observability-and-logging/metrics#about-metrics) and [tracing](https://www.prisma.io/docs/orm/prisma-client/observability-and-logging/opentelemetry-tracing) preview feature to get detailed information about the time spent.

**Reply by teddis** [reply]
@nurul3101 I'm definitely interested in Accelerate and spent the day setting it up and migrating. But I'm running into some problems I can't solve:

1. When I run `yarn generate` or `yarn migrate` followed by `yarn generate`, my prisma/client isn't built, though it outputs no errors. So my build fails. For example, one line of numerous errors:

```
│ src/server.ts:18:12 - error TS2339: Property 'sql' does not exist on type 'typeof Prisma'.
│ 
│ 18     Prisma.sql`WITH user_connections AS (SELECT ${profile_id}                               AS profile_from_id,
```

2. Locally, I only want to use local Postgres and no edge functions, so I'm using the following conditionally to also allow in production:

```
let PrismaClient
if (process.env.NODE_ENV === 'production') {
  PrismaClient = require('@prisma/client/edge').PrismaClient
} else {
  PrismaClient = require('@prisma/client').PrismaClient
}
import {withAccelerate} from '@prisma/extension-accelerate'
```

3. I'm not sure how to properly configure `engine` for local and production/edge. My schema currently has:

```
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["fullTextSearch", "fullTextIndex"]
  engineType      = "library"
}
```

But in production I can't modify the schema.prisma engine field.

I employ package.json script for Vercel: 
```
    "postinstall": "cd ../../packages/prisma && yarn run generate && yarn run deploy"
```

Thanks for your help. Love to get this running ASAP.

**Reply by nurul3101** [reply]
Is there any specific reason you are combining CJS with ESM while instantiating the Prisma Client.
Instead of instantiating the Prisma Client like this:
```ts
let PrismaClient
if (process.env.NODE_ENV === 'production') {
  PrismaClient = require('@prisma/client/edge').PrismaClient
} else {
  PrismaClient = require('@prisma/client').PrismaClient
}
import {withAccelerate} from '@prisma/extension-accelerate'
```
I would recommend you to use the code snippet specified here in this [guide](https://www.prisma.io/docs/accelerate/local-development). This guide shows how you can use local Postgres for local development.

Also, there shouldn't be any need to specify any engineType in `schema.prisma` file, is there any specific reason why you added it?

**Reply by alexey-sh** [reply]
```
import { PrismaClient } from '@prisma/client/edge'
import { PrismaClient } from '@prisma/client'
import Prisma from '@prisma/client';
```

non of above supports `Prisma.sql/PrismaClient.sql` in typescript
tbh looks a bit annoying

**Reply by ludralph** [reply]
Hi @alexey-sh 

Could you please open a new discussion and explain your use case?

**Comment by ludralph** [comment]
Hi @teddis 

Did the suggestion by Nurul help in resolving this issue for you?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
